backend/handler_middleware.go

```diff
@@ -0,0 +1,177 @@
+package backend
+
+import (
+	"context"
+	"errors"
+)
+
+var (
+	errNilRequest = errors.New("req cannot be nil")
+	errNilSender  = errors.New("sender cannot be nil")
+)
+
+// HandlerMiddleware is an interface representing the ability to create a middleware
+// that implements the Handler interface.
+type HandlerMiddleware interface {
+	// CreateHandlerMiddleware creates a new Handler by decorating next Handler.
+	CreateHandlerMiddleware(next Handler) Handler
+}
+
+// The HandlerMiddlewareFunc type is an adapter to allow the use of ordinary
+// functions as HandlerMiddleware's. If f is a function with the appropriate
+// signature, HandlerMiddlewareFunc(f) is a HandlerMiddleware that calls f.
+type HandlerMiddlewareFunc func(next Handler) Handler
+
+// CreateHandlerMiddleware implements the HandlerMiddleware interface.
+func (fn HandlerMiddlewareFunc) CreateHandlerMiddleware(next Handler) Handler {
+	return fn(next)
+}
+
+// MiddlewareHandler decorates a Handler with HandlerMiddleware's.
+type MiddlewareHandler struct {
+	handler Handler
+}
+
+// HandlerFromMiddlewares creates a new MiddlewareHandler implementing Handler that decorates finalHandler with middlewares.
+func HandlerFromMiddlewares(finalHandler Handler, middlewares ...HandlerMiddleware) (*MiddlewareHandler, error) {
+	if finalHandler == nil {
+		return nil, errors.New("finalHandler cannot be nil")
+	}
+
+	return &MiddlewareHandler{
+		handler: handlerFromMiddlewares(middlewares, finalHandler),
```

Comment:
Here's one rather big difference compared to https://github.com/grafana/grafana/blob/6a3dbe7d419ea7b3a9048169c819a16fd8082b47/pkg/plugins/manager/client/decorator.go that re-creates the middleware + final handler chain on each request. 

I don't think this should matter given no middlewares keep state 👀 Could be changed otherwise

Reply:
Performance improvements I would imagine when not having to recreate over and over again. This is just a guess so I rhink I'll change to how it was as a first step

Reply:
up to you really, doesn't seem like a troublesome change.

Reply:
I reverted to make it work like earlier, creating the handler with middlewares on each invoke